python/lsst/daf/butler/tests/butler_query.py

```diff
@@ -0,0 +1,1228 @@
+            Test("-tract,-visit", "tract,visit", ((1, 2), (1, 2), (0, 2), (0, 2), (0, 1), (0, 1))),
+            Test(
+                "tract.id,visit.id",
+                "tract,visit",
+                ((0, 1), (0, 1), (0, 2)),
+                limit=(3,),
+            ),
+            Test("-tract,-visit", "tract,visit", ((1, 2), (1, 2), (0, 2)), limit=(3,)),
+            Test("tract,visit", "tract,visit", ((0, 2), (1, 2), (1, 2)), limit=(3, 3)),
+            Test("-tract,-visit", "tract,visit", ((0, 1),), limit=(3, 5)),
+            Test(
+                "tract,visit.exposure_time", "tract,visit", ((0, 2), (0, 2), (0, 1), (0, 1), (1, 2), (1, 2))
+            ),
+            Test(
+                "-tract,-visit.exposure_time", "tract,visit", ((1, 2), (1, 2), (0, 1), (0, 1), (0, 2), (0, 2))
+            ),
+            Test("tract,-exposure_time", "tract,visit", ((0, 1), (0, 1), (0, 2), (0, 2), (1, 2), (1, 2))),
+            Test("tract,visit.name", "tract,visit", ((0, 1), (0, 1), (0, 2), (0, 2), (1, 2), (1, 2))),
+            Test(
+                "tract,-timespan.begin,timespan.end",
+                "tract,visit",
+                ((0, 2), (0, 2), (0, 1), (0, 1), (1, 2), (1, 2)),
+            ),
+            Test("visit.day_obs,exposure.day_obs", "visit,exposure", ()),
+            Test("visit.timespan.begin,-exposure.timespan.begin", "visit,exposure", ()),
+            Test(
+                "tract,detector",
+                "tract,detector",
+                ((0, 1), (0, 2), (0, 3), (0, 4), (1, 1), (1, 2), (1, 3), (1, 4)),
+                datasets="flat",
+                collections="imported_r",
+            ),
+            Test(
+                "tract,detector.full_name",
+                "tract,detector",
+                ((0, 1), (0, 2), (0, 3), (0, 4), (1, 1), (1, 2), (1, 3), (1, 4)),
+                datasets="flat",
+                collections="imported_r",
+            ),
+            Test(
+                "tract,detector.raft,detector.name_in_raft",
+                "tract,detector",
+                ((0, 1), (0, 2), (0, 3), (0, 4), (1, 1), (1, 2), (1, 3), (1, 4)),
+                datasets="flat",
+                collections="imported_r",
+            ),
+        )
+
+        for test in test_data:
+            order_by = test.order_by.split(",")
+            keys = test.keys.split(",")
+            query = do_query(keys, test.datasets, test.collections).order_by(*order_by)
+            if test.limit is not None:
+                query = query.limit(*test.limit)
+            dataIds = tuple(tuple(dataId[k] for k in keys) for dataId in query)
+            self.assertEqual(dataIds, test.result)
+
+            # and materialize
+            query = do_query(keys).order_by(*order_by)
+            if test.limit is not None:
+                query = query.limit(*test.limit)
+            with self.assertRaises(RelationalAlgebraError):
+                with query.materialize():
+                    pass
```

Comment:
Making this
```
with query.materialize(): pass
```
seems like it'd silence the spurious code coverage issue here.

Reply:
Hmm, `black` is not happy about inlining.

Reply:
I am not surprised.

Reply:
You can always do

```python
pass  # pragma: no cover
```
I think.